lib/tapioca/dsl/helpers/active_record_column_type_helper.rb

```diff
@@ -114,7 +114,10 @@ def column_type_for(column_name)
             end
 
           if @column_type_option.persisted? && !column&.null
-            [getter_type, setter_type]
+            # It's possible that when ActiveModel::Type::Value is used, the signature being reflected on in
+            # ActiveModelTypeHelper.type_for(type_value) may say the type can be nilable. However, if the type is
+            # persisted and the column is not nullable, we can assume it's not nilable.
+            [as_non_nilable_type(getter_type), as_non_nilable_type(setter_type)]
```

Comment:
I'm wary of adding this overhead to every type by default to cover for an edge case. I'd like to understand why changing the signature of `deserialize` and [other methods](https://github.com/Shopify/tapioca/blob/a4ec8e1260c7c841ca2422f5736cfa710d9f78b0/lib/tapioca/dsl/helpers/active_model_type_helper.rb#L17-L20) is not an acceptable solution.

Reply:
Sure thing. So we use custom types pretty widely both for `id` columns and as nilable foreign keys. I'm not sure if there's a way around it for the use case where the type can be used both for nullable and non-nullable columns:

```ruby
sig { params(value: T.nilable(Numeric)).returns(T.nilable(::CustomType))}
def deserialize(value)
  CustomType.new(value) if value
end
```

I could look at moving the call into just the branch of `type_for_activerecord_value` that covers this case. I think it would just mean needing to pass in `@column_type_option` and `column` into that method.

Reply:
(I guess just column since the other is an instance variable)